src/theory/strings/theory_strings_rewriter.cpp

```diff
@@ -1859,9 +1859,55 @@ Node TheoryStringsRewriter::rewriteIndexof( Node node ) {
     }
   }
 
-  Node len0 = nm->mkNode(kind::STRING_LENGTH, node[0]);
-  Node len1 = nm->mkNode(kind::STRING_LENGTH, node[1]);
-  Node len0m2 = nm->mkNode(kind::MINUS, len0, node[2]);
+  if (node[0] == node[1])
+  {
+    // indexof( x, x, 0 ) --> 0
+    if (node[2].isConst())
+    {
+      if (node[2].getConst<Rational>().sgn() == 0)
+      {
+        Node negone = nm->mkConst(Rational(0));
+        return returnRewrite(node, negone, "idof-eq-cst-start");
+      }
+      else
+      {
+        Assert(node[2].getConst<Rational>().sgn() == 1);
```

Comment:
Actually, this case is covered by `checkEntailArith(node[2], true)`, could just make the if statement on line 1865 `node[2].isConst() && node[2].getConst<Rational>().sgn() == 0`.